HeaderParsing README (#4528)
dotnet/extensions: src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md

# Microsoft.AspNetCore.HeaderParsing

This package provides services to for strongly typed header parsing and value caching.

In particular:

- There are many different headers with subtle format differences that are hard to parse correctly in all cases. This package provides predefined parsers for commonly used headers.

> - Your application may parse the same header value multiple times per request. This package automatically caches the parsed values of these headers to make your application faster.

Reviewer: ```suggestion
- Your application may parse the same header value multiple times per request. These APIs automatically cache the parsed values of these headers to make your application faster.
```

- [ContentType](https://learn.microsoft.com/dotnet/api/microsoft.net.http.headers.headernames.ContentType)
- [Cookie](https://learn.microsoft.com/dotnet/api/microsoft.net.http.headers.headernames.Cookie)
- [Date](https://learn.microsoft.com/dotnet/api/microsoft.net.http.headers.headernames.Date)
- [IfMatch](https://learn.microsoft.com/dotnet/api/microsoft.net.http.headers.headernames.IfMatch)
- [IfModifiedSince](https://learn.microsoft.com/dotnet/api/microsoft.net.http.headers.headernames.IfModifiedSince)
- [IfNoneMatch](https://learn.microsoft.com/dotnet/api/microsoft.net.http.headers.headernames.IfNoneMatch)
- [IfRange](https://learn.microsoft.com/dotnet/api/microsoft.net.http.headers.headernames.IfRange)
- [IfUnmodifiedSince](https://learn.microsoft.com/dotnet/api/microsoft.net.http.headers.headernames.IfUnmodifiedSince)
- [Range](https://learn.microsoft.com/dotnet/api/microsoft.net.http.headers.headernames.Range)
- [Referer](https://learn.microsoft.com/dotnet/api/microsoft.net.http.headers.headernames.Referer)
- XForwardedFor (`X-Forwarded-For`)

## Options

The `HeaderParsingOptions` class is used to configure common behaviors for this feature.

> - In case you configured parsing of a specific header, but the header is not present in the current HTTP request, you can pre-configure default values to be returned in such cases. See the `DefaultValues` option.

Reviewer 2: ```suggestion
- You can pre-configure default values to be returned when the current HTTP request contains headers for which you have not configured parsing. See the `DefaultValues` option.
```
?

> - The number of cached values can be changed per header using the `MaxCachedValuesPerHeader` option. If not configured the `DefaultMaxCachedValuesPerHeader` value will be used.

Reviewer 2: nit, you're mixing active ("you can") and passive ("can be")

```suggestion
- You can change the number of cached values using the `MaxCachedValuesPerHeader` option. If not configured the `DefaultMaxCachedValuesPerHeader` value will be used.
```

## Metrics

The package generates the following metrics:

| Metric name | Metric value | Dimension | Dimension value |
|-|-|-|-|
| `HeaderParsing.ParsingErrors` | Increased by **1** with every parsing error     | `HeaderName` | Header name                                       |
|                               |                                                 | `Kind`       | Error message returned by the `TryParse()` method |
| `HeaderParsing.CacheAccess`   | Increased by **1** with every cache hit or miss | `HeaderName` | Header name                                       |
|                               |                                                 | `Type`       | Either `Hit` or `Miss` for the internal header cache hits and misses respectively |

## Custom headers parsing

Although this package provides support for parsing many HTTP headers, you sometimes need to parse custom headers.
You can easily extend the header parsing model to support your custom headers by implementing the abstract class `HeaderParser` and then register an 
instance of this class in the `IHeaderRegistry` interface.

## Feedback & Contributing

> For any feedback or contributions, please visit us in [our GitHub repo](https://github.com/dotnet/extensions).

Reviewer 2: I kind of liked the "We welcome.." wording I suggested in Chris's PR, because it actively encourages folks to engage not just telling them where ...

Author: ```suggestion
We welcome feedback and contributions in [our GitHub repo](https://github.com/dotnet/extensions).
```

# Microsoft.AspNetCore.HeaderParsing

This package provides services to for strongly typed header parsing and value caching.

In particular:

- There are many different headers with subtle format differences that are hard to parse correctly in all cases. This package provides predefined parsers for commonly used headers.
- Your application may parse the same header value multiple times per request. This package automatically caches the parsed values of these headers to make your application faster.
- This package also provides logging and metering of bad header values.

> ## Getting started
> 
> ### Install the package

Reviewer: Seems redundant?
```suggestion
## Install the package
```

Author: Followed @joperezr's README for consistency.

Reviewer: Where?

Author: https://github.com/dotnet/extensions/pull/4512/files#diff-358a8dab66534c16517c7dc752a9f3926e6a3b4b3c2b713cd32d908d8be1f73bR5-R7

From the command-line:

```dotnetcli
dotnet add package Microsoft.AspNetCore.HeaderParsing
```

Or directly in the C# project file:

```xml
<ItemGroup>

>   <PackageReference Include="Microsoft.AspNetCore.HeaderParsing" Version="*" />

Reviewer: Don't encourage bad habits
```suggestion
  <PackageReference Include="Microsoft.AspNetCore.HeaderParsing" Version="[CURRENTVERSION]" />
```

Author: I had no better idea :/ I noticed that @joperezr cheated and only mentioned the dotnet-cli way ;) I think we should converge and use the same pattern everywhere. At least this starts the discussion.

</ItemGroup>
```
## Usage Example

### Registering the services

The services can be registered using one of the `AddHeaderParsing` overloads:

```csharp
public static IServiceCollection AddHeaderParsing(this IServiceCollection services);

> public static IServiceCollection AddHeaderParsing(this IServiceCollection services, Action<HeaderParsingOptions> configuration);

Reviewer: Shouldn't this be 
```suggestion
public static IServiceCollection AddHeaderParsing(this IServiceCollection services, Action<HeaderParsingOptions> configure);
```

Author: Thanks, I will report it to the API review. And the other instances I found BTW.

Author: I am merging this suggestion, but the API is wrong, this will need to be fixed.

Reviewer: Want to just fix the API in this PR?

Author: Done
